keps/sig-auth/4317-pod-certificates/README.md

```diff
@@ -0,0 +1,1313 @@
+	// and must be set at the same time as certificateChain.  Once populated,
+	// this field is immutable.
+  //
+  // This field is only a hint.  Kubelet may start refreshing before or after
+  // this time if necessary.
+	//
+	// +optional
+	BeginRefreshAt *metav1.Time `json:"beginRefreshAt,omitempty" protobuf:"bytes,5,opt,name=beginRefreshAt"`
+
+	// notAfter is the time at which the certificate expires.  This field is set
+	// via the /status subresource.  Once populated, it is immutable.  The 
+  // signer must set this field at the same time it sets certificateChain.
+	//
+	// +optional
+	NotAfter *metav1.Time `json:"notAfter,omitempty" protobuf:"bytes,6,opt,name=notAfter"`
+}
+
+// Well-known condition types for PodCertificateRequests
+const (
+	// Denied indicates the request was denied by the signer.
+	PodCertificateRequestConditionTypeDenied string = "Denied"
+	// Failed indicates the signer failed to issue the certificate.
+	PodCertificateRequestConditionTypeFailed string = "Failed"
+	// SuggestedKeyType is an auxiliary condition that a signer can attach if it
+	// denied the request due to an unsupported key type.
+	PodCertificateRequestConditionTypeSuggestedKeyType string = "SuggestedKeyType"
+)
+
+// Well-known condition reasons for PodCertificateRequests
+const (
+	// UnsupportedKeyType should be set on "Denied" conditions when the signer
+	// doesn't support the key type of publicKey.
+	PodCertificateRequestConditionUnsupportedKeyType string = "UnsupportedKeyType"
+)
+```
+
+### PodCertificate Projected Volume Sources
+
+The PodCertificate projected volume source instructs kubelet to inject a
+credential bundle consisting of a private key and certificate chain into the
+container fileystem.  The kubelet handles generating the key, requesting a
+certificate from the named signer, and renewing the certificate before it
+expires.
+
+The kubelet supports only a few key generation strategies, each named by a
+string that encapsulates both the key type and any parameters necessary (for
+example, RSA modulus size).  The intention is to offer a tasting menu of
+reasonable key choices, rather than offering a flexibility to a wide variety of
+parameters. Key types supported by kubelet are "RSA3072", "RSA4096",
+"ECDSAP256", "ECDSAP384", and "ED25519".  If no key type is specified, kubelet
+defaults to "ECDSAP256".
+
+Once the key is generated, kubelet creates a PodCertificateRequest with the details of the pod that is mounting the volume.  The kubelet then waits for the PodCertificateRequest to be issued, denied, or failed.
+
+If the PodCertificateRequest is issued, then the kubelet writes both the generated private key and the issued certificate chain into the pod filesystem as a single file with multiple PEM blocks.  The first block in the PEM file contains the private key, and the remaining blocks contain the certificate chain in the same order they were in the PodCertificateRequest.
```

Comment:
I am a little concerned with storing the privkey inline with the cert chain? I feel like that just promotes people sending around this single file with the cert chain and maybe not realizing it also contains the private key.

Might be better to separate the private key and chain into explicitly different files.

Reply:
This is an option in the configuration --- the pod author can choose a single file, or separate files for key and certificates.